daemon/cmd/policy.go

```diff
@@ -363,12 +343,9 @@ func (d *Daemon) policyAdd(sourceRules policyAPI.Rules, opts *policy.AddOptions,
 		metrics.PolicyImplementationDelay.WithLabelValues(source).Observe(duration.Seconds())
 	})
 
-	// remove prefixes of replaced rules above. Refcounts have been incremented
-	// above, so any decrements here will be no-ops for CIDRs that are re-added,
-	// and will trigger deletions for those that are no longer used.
+	// Remove prefixes of replaced rules above from the d.prefixLengths tracker.
 	if len(removedPrefixes) > 0 {
 		logger.WithField("prefixes", removedPrefixes).Debug("Decrementing replaced CIDR refcounts when adding rules")
```

Comment:
I see your thinking, but this is related with `Daemon.prefixLengths`, nothing to do with ipcache. This corresponds to the `counter.PrefixLengthCounter` logic where unique prefix lengths are reference-counted (old logic for LPM tracking support on v4.9, which is now no longer necessary). On the other hand, I think you're concerned about reference counting on local ("CIDR") identities. Any such reference counting logic should likely be tracked inside IPCache since IPCache is the one doing the allocation and release of those local identity references.

Off the top of my head, I don't think we have debug messages for ipcache identity alloc/release, but personally I find the debug logs way too verbose to be useful so I just don't pay attention to them.

The right solution here is to delete the prefixLengths logic entirely. That item is tracked in https://github.com/cilium/cilium/issues/22116 .

Reply:
Sounds reasonable to me.